File: host/src/tts/session.ts
import { FrameEncoder } from './frame.ts'
import type { AudioEvent } from './schema.ts'

/** Stand-in speaking rate. A real engine derives frame count from the model. */
const MS_PER_CHAR = 60

/** Unicode White_Space: where one spoken word ends and the next begins. */
const WHITESPACE = /[\t\n\v\f\r \u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]/

/** One speaking session: text tokens in, audio frames out, word by word. */
export class Session {
  #encoder: FrameEncoder
  #speed: number
  #wordChars = 0

  constructor(sampleRate: number, speed: number) {
    this.#encoder = new FrameEncoder(sampleRate)
    this.#speed = speed
  }

  /**
   * Feed a text token (an LLM token, a word, a whole line: any slice). Returns
   * the frames that are ready, which is why speech starts long before the
   * sentence is complete.
   */
  pushText(text: string): AudioEvent[] {
    for (const ch of text) {
      if (WHITESPACE.test(ch)) this.#closeWord(1)
      else this.#wordChars += 1
    }
    return this.#encoder.drain()
  }

  /**
   * End the utterance: flushes the trailing audio, emits exactly one `end`
   * event, and resets the session for the next line.
   */
  finish(): AudioEvent[] {
    this.#closeWord(0)
    const out: AudioEvent[] = this.#encoder.flush()
    out.push({ type: 'end', durationMs: this.#encoder.spokenMs() })
    this.#encoder.reset()
    return out
  }

  #closeWord(gapChars: number): void {
    if (this.#wordChars === 0) return
    const chars = this.#wordChars + gapChars
    this.#encoder.owe(this.#encoder.samplesForMs((chars * MS_PER_CHAR) / this.#speed))
    this.#wordChars = 0
  }
}

File: tts/src/session.rs
//! One speaking session: text tokens in, audio frames out, word by word.

use crate::frame::FrameEncoder;
use crate::schema;
use serde_json::{json, Value};

/// Stand-in speaking rate. A real engine derives frame count from the model.
const MS_PER_CHAR: f64 = 60.0;

pub struct Session {
    encoder: FrameEncoder,
    speed: f64,
    word_chars: u64,
}

impl Session {
    pub(crate) fn new(sample_rate: u32, speed: f64) -> Self {
        Session { encoder: FrameEncoder::new(sample_rate), speed, word_chars: 0 }
    }

    /// Feed a text token (an LLM token, a word, a whole line: any slice).
    /// Returns the audio frames that are ready, which is why speech starts
    /// long before the sentence is complete.
    pub fn push_text(&mut self, text: &str) -> Vec<Value> {
        for ch in text.chars() {
            if ch.is_whitespace() {
                self.close_word(1);
            } else {
                self.word_chars += 1;
            }
        }
        self.encoder.drain()
    }

    /// End the utterance: flushes the trailing audio, emits exactly one `end`
    /// event, and resets the session for the next line.
    pub fn finish(&mut self) -> Vec<Value> {
        self.close_word(0);
        let mut out = self.encoder.flush();
        let end = json!({"type": "end", "durationMs": self.encoder.spoken_ms()});
        debug_assert!(schema::event().is_valid(&end), "end off-contract: {end}");
        out.push(end);
        self.encoder.reset();
        out
    }

    fn close_word(&mut self, gap_chars: u64) {
        if self.word_chars == 0 {
            return;
        }
        let chars = (self.word_chars + gap_chars) as f64;
        let samples = self.encoder.samples_for_ms(chars * MS_PER_CHAR / self.speed);
        self.encoder.owe(samples);
        self.word_chars = 0;
    }
}
